dev/core/src/com/google/gwt/dev/jjs/impl/ImplementRecordComponents.java

```diff
@@ -0,0 +1,231 @@
+    body.getBlock().addStmt(fieldReference.makeReturnStatement());
+    method.setBody(body);
+  }
+
+  private void implementHashCode(JRecordType type, JMethod method, SourceInfo info) {
+    final JExpression hashcodeStatement;
+    if (type.getFields().isEmpty()) {
+      // No fields, just emit hashcode=0
+      hashcodeStatement = new JIntLiteral(info, 0);
+    } else {
+      List<JExpression> exprs = Lists.newArrayListWithCapacity(type.getFields().size());
+      for (JField field : type.getFields()) {
+        JFieldRef jFieldRef = new JFieldRef(info, new JThisRef(info, type), field, type);
+        if (jFieldRef.getType().isPrimitiveType()) {
+          exprs.add(autoboxUtils.box(jFieldRef, (JPrimitiveType) jFieldRef.getType()));
+        } else {
+          exprs.add(jFieldRef);
+        }
+      }
+      JMethod hash = program.getIndexedMethod(RuntimeConstants.OBJECTS_HASH);
+      hashcodeStatement = new JMethodCall(info, null, hash, exprs);
+    }
+    JMethodBody body = new JMethodBody(info);
+    body.getBlock().addStmt(hashcodeStatement.makeReturnStatement());
+
+    method.setBody(body);
+  }
+
+  private void implementEquals(JRecordType type, JMethod method, SourceInfo info) {
+    JMethodBody body = new JMethodBody(info);
+    JParameter otherParam = method.getParams().get(0);
+
+    // Equals is built from first == check between this and other, as a fast path for the same
+    // object and also to ensure that other isn't null. Then MyRecord.class == other.getClass(),
+    // and now we know they're the same type and can cast safely to access fields for the rest.
+    JBinaryOperation eq =
+            new JBinaryOperation(info, JPrimitiveType.BOOLEAN, JBinaryOperator.EQ,
+                    new JThisRef(info, type),
+                    otherParam.createRef(info));
+    body.getBlock().addStmt(new JIfStatement(info, eq,
+            JBooleanLiteral.TRUE.makeReturnStatement(), null));
+
+    JBinaryOperation sameTypeCheck =
+            new JBinaryOperation(info, JPrimitiveType.BOOLEAN, JBinaryOperator.NEQ,
+                    new JClassLiteral(info, type),
+                    new JMethodCall(info, otherParam.createRef(info), getClassMethod));
+    body.getBlock().addStmt(new JIfStatement(info, sameTypeCheck,
+            JBooleanLiteral.FALSE.makeReturnStatement(), null));
+
+    // Create a local to assign to and compare each component
+    JLocal typedOther = JProgram.createLocal(info, "other", type, true, body);
+    // We can use an unsafe cast since we know the check will succeed
+    JUnsafeTypeCoercion uncheckedCast =
+            new JUnsafeTypeCoercion(info, type, otherParam.createRef(info));
+    JBinaryOperation uncheckedAssign = new JBinaryOperation(info, type, JBinaryOperator.ASG,
+            typedOther.createRef(info), uncheckedCast);
+    body.getBlock().addStmt(uncheckedAssign.makeStatement());
+
+    JExpression componentCheck = JBooleanLiteral.TRUE;
+    for (JField field : type.getFields()) {
```

Comment:
This creates an `equals` method that uses `==` for all fields. It will fail for reference types that are equal. You should generate `Objects.equals(this.a, other.a)` for reference types and `PrimitiveWrapper.compare(this.a, other.a) == 0` for primitive types. This is mentioned in the implementation requirements section in the [javadoc of Record.equals()](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Record.html#equals(java.lang.Object)).

Current tests don't catch it because all records being used in tests do not use reference types. Below I added a comment to introduce `RecordWithReferenceType(TopLevelRecord){}`. With the current `equals` implementation a test using this record should fail.

Reply:
Eesh, good catch.